### ISSUE TITLE
Error: Call to a member function instanceOfStorage() on null

### ISSUE THREAD

**Issue by jungleran** [nextcloud/server#10316]
### Steps to reproduce

`sudo -u www-data php occ files:scan --all`

### Expected behaviour

No exception.

### Actual behaviour

An unhandled exception has been thrown:
Error: Call to a member function instanceOfStorage() on null in /var/www/html/apps/files_sharing/lib/Cache.php:81

### Server configuration

Official docker image: nextcloud:13.0.4-apache

Related PR: Fix Error: Call to a member function instanceOfStorage() on null #10315

**Comment by timm2k** [comment]
I had same error after migrating from OC to NC for a single user a month ago. As workaround I've removed ```'share_folder' => '/shared',``` from my config.php. Unfortuantly I'm not able to reproduce this situation now.

**Comment by timm2k** [comment]
I've found a log entry in my backups:
```
{"reqId":"bSUnHnhq23xgyRkQO6eQ","level":3,"time":"2018-08-01T14:39:00+02:00","remoteAddr":"134.106.124.196","user":"lupp5414","app":"remote","method":"PROPFIND","url":"\/remote.php\/webdav\/Shared","message":"Exception: {\"Exception\":\"Error\",\"Message\":\"Call to a member function instanceOfStorage() on null\",\"Code\":0,\"Trace\":\"#0 \\\/var\\\/www\\\/html\\\/nextcloud\\\/lib\\\/private\\\/Files\\\/View.php(2034): OC\\\\Files\\\\View->lockPath('\\\/Shared', 1, false)\\n#1 \\\/var\\\/www\\\/html\\\/nextcloud\\\/lib\\\/private\\\/Files\\\/View.php(1134): OC\\\\Files\\\\View->lockFile('\\\/Shared', 1)\\n#2 \\\/var\\\/www\\\/html\\\/nextcloud\\\/lib\\\/private\\\/Files\\\/View.php(267): OC\\\\Files\\\\View->basicOperation('mkdir', '\\\/Shared', Array)\\n#3 \\\/var\\\/www\\\/html\\\/nextcloud\\\/apps\\\/files_sharing\\\/lib\\\/Helper.php(259): OC\\\\Files\\\\View->mkdir('\\\/\\\/Shared')\\n#4 \\\/var\\\/www\\\/html\\\/nextcloud\\\/apps\\\/files_sharing\\\/lib\\\/SharedMount.php(93): OCA\\\\Files_Sharing\\\\Helper::getShareFolder(Object(OC\\\\Files\\\\View))\\n#5 \\\/var\\\/www\\\/html\\\/nextcloud\\\/apps\\\/files_sharing\\\/lib\\\/SharedMount.php(74): OCA\\\\Files_Sharing\\\\SharedMount->verifyMountPoint(Object(OC\\\\Share20\\\\Share), Array)\\n#6 \\\/var\\\/www\\\/html\\\/nextcloud\\\/apps\\\/files_sharing\\\/lib\\\/MountProvider.php(98): OCA\\\\Files_Sharing\\\\SharedMount->__construct('\\\\\\\\OCA\\\\\\\\Files_Shar...', Array, Array, Object(OC\\\\Files\\\\Storage\\\\StorageFactory))\\n#7 \\\/var\\\/www\\\/html\\\/nextcloud\\\/lib\\\/private\\\/Files\\\/Config\\\/MountProviderCollection.php(108): OCA\\\\Files_Sharing\\\\MountProvider->getMountsForUser(Object(OC\\\\User\\\\User), Object(OC\\\\Files\\\\Storage\\\\StorageFactory))\\n#8 \\\/var\\\/www\\\/html\\\/nextcloud\\\/lib\\\/private\\\/Files\\\/Filesystem.php(445): OC\\\\Files\\\\Config\\\\MountProviderCollection->addMountForUser(Object(OC\\\\User\\\\User), Object(OC\\\\Files\\\\Mount\\\\Manager))\\n#9 \\\/var\\\/www\\\/html\\\/nextcloud\\\/apps\\\/files_sharing\\\/lib\\\/SharedStorage.php(118): OC\\\\Files\\\\Filesystem::initMountPoints('noau7032')\\n#10 \\\/var\\\/www\\\/html\\\/nextcloud\\\/apps\\\/files_sharing\\\/lib\\\/SharedStorage.php(464): OCA\\\\Files_Sharing\\\\SharedStorage->init()\\n#11 \\\/var\\\/www\\\/html\\\/nextcloud\\\/lib\\\/private\\\/Files\\\/Storage\\\/Wrapper\\\/Wrapper.php(494): OCA\\\\Files_Sharing\\\\SharedStorage->getWrapperStorage()\\n#12 \\\/var\\\/www\\\/html\\\/nextcloud\\\/apps\\\/files_sharing\\\/lib\\\/SharedStorage.php(157): OC\\\\Files\\\\Storage\\\\Wrapper\\\\Wrapper->instanceOfStorage('\\\\\\\\OCP\\\\\\\\Files\\\\\\\\IHom...')\\n#13 \\\/var\\\/www\\\/html\\\/nextcloud\\\/lib\\\/private\\\/Files\\\/Storage\\\/Wrapper\\\/Wrapper.php(494): OCA\\\\Files_Sharing\\\\SharedStorage->instanceOfStorage('\\\\\\\\OCP\\\\\\\\Files\\\\\\\\IHom...')\\n#14 \\\/var\\\/www\\\/html\\\/nextcloud\\\/lib\\\/private\\\/Files\\\/FileInfo.php(313): OC\\\\Files\\\\Storage\\\\Wrapper\\\\Wrapper->instanceOfStorage('\\\\\\\\OCP\\\\\\\\Files\\\\\\\\IHom...')\\n#15 \\\/var\\\/www\\\/html\\\/nextcloud\\\/apps\\\/dav\\\/lib\\\/Connector\\\/Sabre\\\/Node.php(301): OC\\\\Files\\\\FileInfo->isMounted()\\n#16 \\\/var\\\/www\\\/html\\\/nextcloud\\\/apps\\\/dav\\\/lib\\\/Connector\\\/Sabre\\\/FilesPlugin.php(307): OCA\\\\DAV\\\\Connector\\\\Sabre\\\\Node->getDavPermissions()\\n#17 \\\/var\\\/www\\\/html\\\/nextcloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/PropFind.php(98): OCA\\\\DAV\\\\Connector\\\\Sabre\\\\FilesPlugin->OCA\\\\DAV\\\\Connector\\\\Sabre\\\\{closure}(*** sensitive parameters replaced ***)\\n#18 \\\/var\\\/www\\\/html\\\/nextcloud\\\/apps\\\/dav\\\/lib\\\/Connector\\\/Sabre\\\/FilesPlugin.php(313): Sabre\\\\DAV\\\\PropFind->handle('{http:\\\/\\\/ownclou...', Object(Closure))\\n#19 [internal function]: OCA\\\\DAV\\\\Connector\\\\Sabre\\\\FilesPlugin->handleGetProperties(Object(Sabre\\\\DAV\\\\PropFind), Object(OCA\\\\DAV\\\\Connector\\\\Sabre\\\\Directory))\\n#20 \\\/var\\\/www\\\/html\\\/nextcloud\\\/3rdparty\\\/sabre\\\/event\\\/lib\\\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\\n#21 \\\/var\\\/www\\\/html\\\/nextcloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(1059): Sabre\\\\Event\\\\EventEmitter->emit('propFind', Array)\\n#22 \\\/var\\\/www\\\/html\\\/nextcloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(981): Sabre\\\\DAV\\\\Server->getPropertiesByNode(Object(Sabre\\\\DAV\\\\PropFind), Object(OCA\\\\DAV\\\\Connector\\\\Sabre\\\\Directory))\\n#23 \\\/var\\\/www\\\/html\\\/nextcloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(1666): Sabre\\\\DAV\\\\Server->getPropertiesIteratorForPath('Shared', Array, 1)\\n#24 \\\/var\\\/www\\\/html\\\/nextcloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/CorePlugin.php(359): Sabre\\\\DAV\\\\Server->generateMultiStatus(Object(Generator), false)\\n#25 [internal function]: Sabre\\\\DAV\\\\CorePlugin->httpPropFind(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#26 \\\/var\\\/www\\\/html\\\/nextcloud\\\/3rdparty\\\/sabre\\\/event\\\/lib\\\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\\n#27 \\\/var\\\/www\\\/html\\\/nextcloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(479): Sabre\\\\Event\\\\EventEmitter->emit('method:PROPFIND', Array)\\n#28 \\\/var\\\/www\\\/html\\\/nextcloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(254): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#29 \\\/var\\\/www\\\/html\\\/nextcloud\\\/apps\\\/dav\\\/appinfo\\\/v1\\\/webdav.php(80): Sabre\\\\DAV\\\\Server->exec()\\n#30 \\\/var\\\/www\\\/html\\\/nextcloud\\\/remote.php(164): require_once('\\\/var\\\/www\\\/html\\\/n...')\\n#31 {main}\",\"File\":\"\\\/var\\\/www\\\/html\\\/nextcloud\\\/lib\\\/private\\\/Files\\\/View.php\",\"Line\":1920}","userAgent":"Mozilla\/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit\/537.36 (KHTML, like Gecko) Chrome\/67.0.3396.87 Safari\/537.36 OPR\/54.0.2952.64","version":"13.0.5.2"}
```

**Comment by MorrisJobke** [comment]
Could we ask you to apply following patch to properly trace the issue:

```patch
diff --git a/apps/files_sharing/lib/Cache.php b/apps/files_sharing/lib/Cache.php
index 352001ecbd..0b3df28a7a 100644
--- a/apps/files_sharing/lib/Cache.php
+++ b/apps/files_sharing/lib/Cache.php
@@ -78,6 +78,9 @@ class Cache extends CacheJail {
 			// in the case where a folder is shared from a Jail we need to ensure that the share Jail
 			// has it's root set relative to the source Jail
 			$currentStorage = $this->storage->getSourceStorage();
+			if ($currentStorage === null) {
+				\OC::$server->getLogger()->logException(new \Exception('currentStorage should not be null but it is'));
+			}
 			if ($currentStorage->instanceOfStorage(Jail::class)) {
 				/** @var Jail $currentStorage */
 				$absoluteRoot = $currentStorage->getJailedPath($absoluteRoot);
```

**Comment by timm2k** [comment]
I've applied this additional logging to our production environment. I'll monitor it for a week and provide a feedback.

**Comment by MorrisJobke** [comment]
@rullzer @icewind1991 Was this the ticket you were discussing in IRC about?

**Comment by MorrisJobke** [comment]
The stack trace looks like this:

```
{
    "reqId": "3gyqIgh7NAK4Tz4HPNJL",
    "level": 3,
    "time": "...",
    "remoteAddr": "...",
    "user": "...",
    "app": "no app in context",
    "method": "PROPFIND",
    "url": "/remote.php/webdav/",
    "message": {
        "Exception": "Exception",
        "Message": "currentStorage should not be null but it is",
        "Code": 0,
        "Trace"
        #0 lib/private/Files/Cache/Wrapper/CacheJail.php(69): OCA\Files_Sharing\Cache->getRoot()
        #1 lib/private/Files/Cache/Wrapper/CacheJail.php(314): OC\Files\Cache\Wrapper\CacheJail->getJailedPath(NULL)
        #2 lib/private/Files/View.php(1724): OC\Files\Cache\Wrapper\CacheJail->getPathById(4819)
        #3 apps/files_sharing/lib/SharedStorage.php(118): OC\Files\View->getPath(4819)
        #4 apps/files_sharing/lib/SharedStorage.php(458): OCA\Files_Sharing\SharedStorage->init()\\n#5 [internal function]: OCA\Files_Sharing\SharedStorage->getSourceStorage()
        #6 lib/private/Files/Storage/Wrapper/Wrapper.php(501): call_user_func_array(Array, Array)
        #7 apps/files_sharing/lib/Cache.php(94): OC\Files\Storage\Wrapper\Wrapper->__call('getSourceStorag...', Array)
        #8 lib/private/Files/Cache/Wrapper/CacheJail.php(313): OCA\Files_Sharing\Cache->getCache()
        #9 lib/private/Files/View.php(1724): OC\Files\Cache\Wrapper\CacheJail->getPathById(4846)
        #10 apps/files_sharing/lib/SharedStorage.php(118): OC\Files\View->getPath(4846)
        #11 apps/files_sharing/lib/SharedStorage.php(463): OCA\Files_Sharing\SharedStorage->init()
        #12 lib/private/Files/Storage/Wrapper/Wrapper.php(490): OCA\Files_Sharing\SharedStorage->getWrapperStorage()
        #13 apps/files_sharing/lib/SharedStorage.php(156): OC\Files\Storage\Wrapper\Wrapper->instanceOfStorage('\\OCP\\Files\\IHom...')
        #14 lib/private/Files/Storage/Wrapper/Wrapper.php(490): OCA\Files_Sharing\SharedStorage->instanceOfStorage('\\OCP\\Files\\IHom...')
        #15 lib/private/Files/FileInfo.php(310): OC\Files\Storage\Wrapper\Wrapper->instanceOfStorage('\\OCP\\Files\\IHom...')
        #16 apps/dav/lib/Connector/Sabre/Node.php(299): OC\Files\FileInfo->isMounted()
        #17 apps/dav/lib/Connector/Sabre/FilesPlugin.php(312): OCA\DAV\Connector\Sabre\Node->getDavPermissions()
        #18 3rdparty/sabre/dav/lib/DAV/PropFind.php(98): OCA\DAV\Connector\Sabre\FilesPlugin->OCA\DAV\Connector\Sabre\{closure}(*** sensitive parameters replaced ***)
        #19 apps/dav/lib/Connector/Sabre/FilesPlugin.php(318): Sabre\DAV\PropFind->handle('{http://...', Object(Closure))\\n#20 [internal function]: OCA\DAV\Connector\Sabre\FilesPlugin->handleGetProperties(Object(Sabre\DAV\PropFind), Object(OCA\DAV\Connector\Sabre\Directory))
        #21 3rdparty/sabre/event/lib/EventEmitterTrait.php(105): call_user_func_array(Array, Array)
        #22 3rdparty/sabre/dav/lib/DAV/Server.php(1059): Sabre\Event\EventEmitter->emit('propFind', Array)
        #23 3rdparty/sabre/dav/lib/DAV/Server.php(981): Sabre\DAV\Server->getPropertiesByNode(Object(Sabre\DAV\PropFind), Object(OCA\DAV\Connector\Sabre\Directory))
        #24 3rdparty/sabre/dav/lib/DAV/Server.php(1666): Sabre\DAV\Server->getPropertiesIteratorForPath('', Array, 1)
        #25 3rdparty/sabre/dav/lib/DAV/CorePlugin.php(355): Sabre\DAV\Server->generateMultiStatus(Object(Generator), false)\\n#26 [internal function]: Sabre\DAV\CorePlugin->httpPropFind(Object(Sabre\HTTP\Request), Object(Sabre\HTTP\Response))
        #27 3rdparty/sabre/event/lib/EventEmitterTrait.php(105): call_user_func_array(Array, Array)
        #28 3rdparty/sabre/dav/lib/DAV/Server.php(479): Sabre\Event\EventEmitter->emit('method:PROPFIND', Array)
        #29 3rdparty/sabre/dav/lib/DAV/Server.php(254): Sabre\DAV\Server->invokeMethod(Object(Sabre\HTTP\Request), Object(Sabre\HTTP\Response))
        #30 apps/dav/appinfo/v1/webdav.php(76): Sabre\DAV\Server->exec()
        #31 remote.php(162): require_once('/var/www/...')
        #32 {main}",
        "File":"apps/files_sharing/lib/Cache.php",
        "Line": 81
    },
    "userAgent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:62.0) Gecko/20100101 Firefox/62.0",
    "version": "12.0.5.3"
}
```

**Comment by rullzer** [comment]
I think this is fixed https://github.com/nextcloud/server/pull/11862

But this still doesn't solve the issue. As a shared mount can thus be invalid and show up weird.

**Comment by centredecalcul** [comment]
In my case the bug occurs in this case:
- users A and B come from user_ldap
- user A shares a folder with user B
- delete A in LDAP, not in nextcloud
- scan user B

**Comment by MorrisJobke** [comment]
Fixed via #11862 on stable12, 13 and 14.
